### PR TITLE
Update device cache

### PR DESCRIPTION
**Pull request by murrant**
Prepare for cache to be accessible in Laravel code
Also a way to have the "primary" device accessible/shared globally without using php globals

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: LibreNMS/Cache/Device.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    /**
+     * Get a device by hostname
+     *
+     * @param string $hostname
+     * @return \App\Models\Device
+     */
+    public function getByHostname($hostname) : \App\Models\Device
+    {
+        $device_id = $device_id = collect($this->devices)->pluck('device_id', 'hostname')->get($hostname);
```
Comment (Jellyfrog) [inline LibreNMS/Cache/Device.php]: Dual assignment

Comment (murrant) [reply LibreNMS/Cache/Device.php]: haha, oops

File: includes/common.php [hunk 2]
```diff
@@ -470,28 +431,8 @@ function getifdescrbyid($id)
...
-
-    if (isset($cache['devices']['id'][$id]['os'])) {
-        $os = $cache['devices']['id'][$id]['os'];
-    } else {
-        $os = dbFetchCell("SELECT `os` FROM `devices` WHERE `device_id` = ?", array($id));
-    }
-
-    return $os;
+    $device = DeviceCache::getByHostname($hostname);
+    return $device ? $device->device_id : null;
```
Comment (Jellyfrog) [inline includes/common.php]: This test is always true

Comment (murrant) [reply includes/common.php]: ah, yes, that was before I changed the cache to always return a device.

**Comment by SourceDoctor** [comment]
@murrant 
Crash on productiv system and testsystem with this path

> 2019-11-14 23:57:09] production.ERROR: Class 'DeviceCache' not found {"userId":2,"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Class 'DeviceCache' not found at /opt/librenms/includes/html/pages/device.inc.php:20)
[stacktrace]
#0 /opt/librenms/app/Http/Controllers/LegacyController.php(51): require()
#1 [internal function]: App\\Http\\Controllers\\LegacyController->index(Object(Illuminate\\Http\\Request), Object(Illuminate\\Session\\Store), 'device/device=7')
#2 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): call_user_func_array(Array, Array)
#3 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(45): Illuminate\\Routing\\Controller->callAction('index', Array)

**Comment by murrant** [comment]
@SourceDoctor how are you updating?  I can't reproduce that.

**Comment by SourceDoctor** [comment]
simple with calling daily.sh in /opt/librenms

**Comment by Jellyfrog** [comment]
Same for us in production.

```
[2019-11-15 14:05:07] production.ERROR: Class 'DeviceCache' not found {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableErr
or(code: 0): Class 'DeviceCache' not found at /opt/librenms/poller.php:167)
[stacktrace]
#0 {main}
"}
```
```
[2019-11-15 14:06:02] production.ERROR: Class 'DeviceCache' not found {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableErr
or(code: 0): Class 'DeviceCache' not found at /opt/librenms/includes/common.php:464)
[stacktrace]
#0 /opt/librenms/LibreNMS/Alert/RunAlerts.php(85): get_dev_attribs(9)
#1 /opt/librenms/LibreNMS/Alert/RunAlerts.php(276): LibreNMS\\Alert\\RunAlerts->describeAlert(Array)
#2 /opt/librenms/LibreNMS/Alert/RunAlerts.php(481): LibreNMS\\Alert\\RunAlerts->issueAlert(Array)
#3 /opt/librenms/alerts.php(53): LibreNMS\\Alert\\RunAlerts->runAlerts()
#4 {main}
"} 
```
```
[2019-11-15 14:06:20] production.ERROR: Class 'DeviceCache' not found {"userId":12,"exception":"[object] (Symfony\\Component\\Debug\\Exception\\Fatal
ThrowableError(code: 0): Class 'DeviceCache' not found at /opt/librenms/includes/common.php:351)
[stacktrace]
#0 /opt/librenms/includes/html/graphs/graph.inc.php(13): device_by_id_cache('56')
#1 /opt/librenms/html/graph.php(27): require('/opt/librenms/i...')
#2 {main}
"} 
```

**Comment by SourceDoctor** [comment]
@Jellyfrog 
you we're right :-)
clearing cache seems to solve it 
thx to @murrant for fixing the issue

**Comment by murrant** [comment]
Don't cache the config unless you are prepared to deal with problems like that ;)

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-58-release-changelog-november-2019/10224/1

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/check-the-cron-job-to-make-sure-it-is-running-and-using-discovery-wrapper-py/10167/2

**Pull request merged**